crates/payload/basic/src/lib.rs

```diff
@@ -654,13 +702,94 @@ where
     let block_gas_limit: u64 = initialized_block_env.gas_limit.try_into().unwrap_or(u64::MAX);
     let base_fee = initialized_block_env.basefee.to::<u64>();
 
+    #[cfg(feature = "optimism")]
+    let block_gas_limit: u64 = attributes.gas_limit.unwrap_or(block_gas_limit);
+
     let mut executed_txs = Vec::new();
     let mut best_txs = pool.best_transactions_with_base_fee(base_fee);
 
     let mut total_fees = U256::ZERO;
 
     let block_number = initialized_block_env.number.to::<u64>();
 
+    // Deposit transactions are always inserted at the top of the block built by the payload.
+    // These transactions are pre-paid on L1 and have no signature.
+    #[cfg(feature = "optimism")]
+    {
+        for deposit_tx in attributes.transactions {
+            // Check if the job was cancelled, if so we can exit early.
+            if cancel.is_cancelled() {
+                return Ok(BuildOutcome::Cancelled)
+            }
+
+            // Convert the deposit transaction to a [TransactionSignedEcRecovered]. This is
+            // purely for the purposes of utilizing the [tx_env_with_recovered] function,
+            // deposit transactions do not have signatures.
+            let deposit_tx = deposit_tx
+                .clone()
+                .try_into_ecrecovered()
+                .map_err(|_| PayloadBuilderError::DepositTransactionRecoverFailed)?;
+
+            // Configure the environment for the block.
+            let env = Env {
+                cfg: initialized_cfg.clone(),
+                block: initialized_block_env.clone(),
+                tx: tx_env_with_recovered(&deposit_tx),
+            };
+
+            let mut evm = revm::EVM::with_env(env);
+            evm.database(&mut db);
+
+            let ResultAndState { result, state } = match evm.transact() {
+                Ok(res) => res,
+                Err(err) => {
+                    match err {
+                        EVMError::Transaction(err) => {
+                            // if the deposit transaction is invalid, we can skip it
+                            trace!(?err, ?deposit_tx, "skipping invalid deposit transaction");
+                            continue
```

Comment:
Invalid deposits are still included in the block, so users can find and see that their deposit isn't censored for arbitrary reasons, but any error is just turned into a failed receipt-status. This also helps safe-block-consolidation just compare block attributes, rather than having to reprocess deposits to find which should have been included, and which shouldn't have.
I think this type of invalid deposit may happen e.g. if the user tries to move ETH that the sender account doesn't have (and doesn't mint).
I'm not sure if you've to change the code here to guarantee inclusion in the block, but it's worth a comment in the code.